### DISCUSSION TITLE
creating-singleton-named-options-with-ioptionsmonitor/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#276] [Announcements]
# creating-singleton-named-options-with-ioptionsmonitor/

In this post I introduce the IOptionsMonitor interface, and show how you can use it in place of IOptionsSnapshot to create Singleton named options

https://andrewlock.net/creating-singleton-named-options-with-ioptionsmonitor/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 17 2019, 08:00, in reply to [Andy Pryor](https://disqus.com/by/disqus_47DZhFpFaW/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/276#discussioncomment-6054147)</em>

---

<p>You can see an example of OnChange being used in the BatchingLoggerProvider, e.g.: <a href="https://github.com/aspnet/Extensions/blob/55c2346714ed114a4525e37cb8cd15794a2ba824/src/Logging/Logging.AzureAppServices/src/BatchingLoggerProvider.cs#L50" rel="nofollow noopener" title="https://github.com/aspnet/Extensions/blob/55c2346714ed114a4525e37cb8cd15794a2ba824/src/Logging/Logging.AzureAppServices/src/BatchingLoggerProvider.cs#L50">https://github.com/aspnet/Extensions/blob/55c2346714ed114a4525e37cb8cd15794a2ba824/src/Logging/Logging.AzureAppServices/src/BatchingLoggerProvider.cs#L50</a> Hope that helps!</p>

[a comment by an automated account is not shown]

**Comment by JonPSmith** [comment]
Hi Andrew,

Can you confirm what I have found about `IOptions` and when they are updated.

I want to update a json file and within the same action / HTTP request I want to obtain the changed data by using  `IOptions`, specifically the `IOptionsMonitor`. However, my tests show that the `IOptionsMonitor`'s `OnChange` isn't triggered if the file updated in the same HTTP request as the `IOptionsMonitor` is in.

I haven't seen any information about when the `IOptions` updates and assumed it triggers on the change using something like the [FileSystemWatcher](https://www.c-sharpcorner.com/UploadFile/puranindia/filesystemwatcher-in-C-Sharp/), which triggers immediately after the file being changed.

Have you found any documentation / code that says when `IOptionsMonitor`'s `OnChange` is triggered?

Not urgent, but would like to know when `IOptions` are updated.
